Add new open APIs for generic GRIB Fieldsets

alexamici opened the pull request:
Add entry points the generic implementations in #243 that do not assume the storage is a GRIB file.

Closes #243

alexamici commented:
@iainrussell the code for the PR should be ready, I need to add the documentation to the new entry points `open_container` and `open_from_index` and then you should be able to test it.

I'm not sure wether to add a section to the README since this is a developer API, not a user one.

alexamici commented:
@iainrussell and @b8raoult after a huge cleanup and rename of the interface to `Field` and `Fieldset`, I added a short documentation in `ADVANCED_USAGE.rst`.

I think the PR is complete and ready to be tested.

alexamici commented:
Final bit: we tested the performance of the GRIB to netCDF translation with the new code with no regression.

iainrussell commented:
This looks great, @alexamici  - I'm on leave this week and next, so will probably have a proper look at it in the week of 15th November.
Cheers,
Iain

iainrussell commented:
Hi @alexamici,

I started testing this on Friday, and managed to create a new proxy class in Metview that could be passed to this in place of a Fieldset (with more time, I'm sure we can pass an actual Fieldset, I just didn't want to risk changing behaviour at the moment). And it worked, I got a nice dataset from it, which had (most of) the right dimensions and the plot looked reasonable  :)  So well done!

I noticed one thing that I think is not working properly: date&time. The time is always set to 1970. I noticed that the cfgrib function `from_grib_date_time()` is not being called at all. Perhaps this is a point of discussion - the current implementation on the Metview end will return all the normal GRIB keys, e.g. dataDate, dataTime, etc, so the actual time can be computed in the same way as it would be with a normal GRIB file. The alternative is that cfgrib asks for 'time' (as it seems to be), and we return a datetime object - I tried that, but it failed - if this is the preferred way to go, do you have an example of how to specify the date/time in the dict example?

I also had an issue, but it might have been a problem in my environment - most of the cfgrib tests did not initially work if a GRIB file had an index file, but the tests worked once I deleted the index files. It actually seems to be working now, so it may have been VSCode being funny. I got this message: "AttributeError: 'FileIndex' object has no attribute 'field_ids_index'". If that seems impossible, then ignore this, because I do not get that error today! But otherwise it might be good to check.

But this is a great achievement, a major refactoring of the code, allowing developers to 'impersonate' GRIB files with other classes!

Cheers,
Iain

alexamici commented on cfgrib/messages.py:
```diff
@@ -318,43 +308,38 @@ def __len__(self) -> int:
-        count_offsets = {}  # type: T.Dict[int, int]
         header_values_cache = {}  # type: T.Dict[T.Tuple[T.Any, type], T.Any]
-        for offset_field, message in filestream.items():
+        if hasattr(fieldset, 'items') and callable(fieldset.items):
```
@b8raoult `black` forces double quotes for strings. I'll fix it.

alexamici commented on cfgrib/messages.py:
```diff
@@ -318,43 +308,38 @@ def __len__(self) -> int:
+        if hasattr(fieldset, 'items') and callable(fieldset.items):
+            iterate = fieldset.items()
+        else:
+            iterate = enumerate(fieldset)
```
@b8raoult `mypy` gets confused because the two assign to `itarate` are not the same type. I'll fix it.

alexamici commented:
@b8raoult & @iainrussell 

I added around an internal use of `typing.Protocol` around https://github.com/ecmwf/cfgrib/pull/265/commits/62e3a786cb3e9dc420c9a8d7e192a7d67e8bf949 but that is not supported in Python 3.7 and there's no backward compatibility module.

Since this is an internal definition and only useful to clean up the type hints inside the cose, not at API level, I'll try to find another way to get types right, which will probably look a bit uglier.

Feature-wise the current implementation supports mapping and sequence fieldsets with lazy loading.

alexamici commented:
@iainrussell & @b8raoult the PR is now ready for review and testing (and merge).

The new API is supports the sequence-like `Fieldset` and the `MappingFieldset`, both supports lazy-loading and I added basic documentation of the new API.

iainrussell commented:
@alexamici that's great! I think then the only outstanding thing is the generation of computed keys, related to time in particular.

alexamici commented:
@iainrussell I added the `compute_index_keys` to the public API and moved the computed keys machinery to use an adapter instead of a class in all read-only code paths.

That should close the remaining requests AND make the PR a real monster.

On the plus side all tests still work after I sync'ed to the new interfaces and I couldn't measure any performance regression.

It is OK to merge for me.
